llvm/include/llvm/LTO/legacy/ThinLTOCodeGenerator.h

```diff
@@ -125,11 +154,61 @@ class ThinLTOCodeGenerator {
   struct CachingOptions {
     std::string Path;                    // Path to the cache, empty to disable.
     CachePruningPolicy Policy;
+    enum class CacheType {
+      CacheDirectory,
+      CAS,
+      RemoteService,
+    } Type;
+    std::unique_ptr<cas::ObjectStore> CAS;
+    std::unique_ptr<cas::ActionCache> Cache;
+    Optional<cas::remote::ClientServices> Service;
   };
 
   /// Provide a path to a directory where to store the cached files for
   /// incremental build.
-  void setCacheDir(std::string Path) { CacheOptions.Path = std::move(Path); }
+  Error setCacheDir(std::string Path) {
+    // CacheDir can only be set once.
+    if (!CacheOptions.Path.empty())
+      return Error::success();
```

Comment:
The environment variable overwrite happens when codegen is created so it would work even the user didn't specify the cache path through C API. If user did set the cache, it is going to hit the condition and skipped.